test/OpenTelemetry.Instrumentation.AspNetCore.Tests/ExceptionFilterTests.cs

```diff
@@ -0,0 +1,185 @@
+
+            // Arrange
+            using (var client = this.factory
+                .WithWebHostBuilder(builder => builder.ConfigureTestServices(
+                    (s) => ConfigureTestServices(s, 2, ref exportedItems)))
+                .CreateClient())
+            {
+                // Act
+                var response = await client.GetAsync("/api/error");
+
+                WaitForActivityExport(exportedItems, 1);
+            }
+
+            // Assert
+            AssertException(exportedItems);
+        }
+
+        public void Dispose()
+        {
+            this.tracerProvider?.Dispose();
+        }
+
+        private static void WaitForActivityExport(List<Activity> exportedItems, int count)
+        {
+            // We need to let End callback execute as it is executed AFTER response was returned.
+            // In unit tests environment there may be a lot of parallel unit tests executed, so
+            // giving some breezing room for the End callback to complete
+            Assert.True(SpinWait.SpinUntil(
+                () =>
+                {
+                    Thread.Sleep(10);
+                    return exportedItems.Count >= count;
+                },
+                TimeSpan.FromSeconds(1)));
+        }
+
+        private static void ValidateAspNetCoreActivity(Activity activityToValidate, string expectedHttpPath)
+        {
+            Assert.Equal(ActivityKind.Server, activityToValidate.Kind);
+            Assert.Equal(HttpInListener.ActivitySourceName, activityToValidate.Source.Name);
+            Assert.Equal(HttpInListener.Version.ToString(), activityToValidate.Source.Version);
+            Assert.Equal(expectedHttpPath, activityToValidate.GetTagValue(SemanticConventions.AttributeHttpTarget) as string);
+        }
+
+        private void ConfigureTestServices(IServiceCollection services, int mode, ref List<Activity> exportedItems)
+        {
+            switch (mode)
+            {
+                case 1:
+                    services.AddMvc(x => x.Filters.Add<ExceptionFilter1>());
+                    break;
+                case 2:
+                    services.AddMvc(x => x.Filters.Add<ExceptionFilter1>());
+                    services.AddMvc(x => x.Filters.Add<ExceptionFilter2>());
+                    break;
+                default:
+                    break;
+            }
+
+            this.tracerProvider = Sdk.CreateTracerProviderBuilder()
+                .AddAspNetCoreInstrumentation(x => x.RecordException = true)
+                .AddInMemoryExporter(exportedItems)
+                .Build();
+        }
+
+        private void AssertException(List<Activity> exportedItems)
+        {
+            Assert.Single(exportedItems);
+            var activity = exportedItems[0];
+
+            var exMessage = "something's wrong!";
+            Assert.Equal("System.Exception", activity.Events.First().Tags.FirstOrDefault(t => t.Key == SemanticConventions.AttributeExceptionType).Value);
```

Comment:
Adding an `Assert.Single(activity.Events)` ensures we have get a single Event and not doing additional work for repeat invocations of the `OnException` event.